python/lsst/daf/butler/_formatter.py

```diff
@@ -51,12 +60,927 @@
     from ._dataset_ref import DatasetRef
     from ._dataset_type import DatasetType
     from ._storage_class import StorageClass
-    from .dimensions import DataCoordinate
+    from .datastore.cache_manager import AbstractDatastoreCacheManager
 
     # Define a new special type for functions that take "entity"
     Entity: TypeAlias = DatasetType | DatasetRef | StorageClass | str
 
 
+class FileIntegrityError(RuntimeError):
+    """The file metadata is inconsistent with the metadata supplied by
+    the datastore.
+    """
+
+
+class FormatterV2(metaclass=ABCMeta):
+    """Interface for reading and writing datasets using URIs.
+
+    The formatters are associated with a particular `StorageClass`.
```

Comment:
A storage class is associated with one formatter in configuration.  I need to think about how Formatter can say what it supports. The current situation is:

* On read the formatter can see what the requested storage class is and generate that directly.
* On write the datastore does a type conversion before sending the in-memory dataset to the formatter

We would need something like a `FormatterV2.can_accept(in_memory_dataset)` API that would allow a formatter to say it's all under control, bypass the type conversion. Maybe it has to take the storageClass instead. Need to think. It's not solely a case of "can you accept X" it's also an assertion that the formatter can read it back as the registry storage class python type.

Reply:
I have implemented `can_accept` so that @erykoff can prevent conversions to DataFrame on write.